Google draco exporter (.drc)

tentone opened the pull request:
Hi

Added a DracoExporter, using the draco encoder js lib and an example.

The exporter receives some options that can be used to control the compression level and quantification of the attributes.

Right now it is using Geometry as input, can be changed to export entire objects later on (or keep methods for both).

Cheers

tentone committed:
Draco exporter

tentone committed:
Derp

tentone committed:
Derp

donmccurdy commented:
Thanks! For consistency with DRACOLoader, do you mind renaming this to `DRACOExporter`?

> Right now it is using Geometry as input, can be changed to export entire objects later on (or keep methods for both).

BufferGeometry as input would be ideal – we're gradually converting all examples from Geometry to BufferGeometry. I wouldn't advise supporting Scene, Object3D, or Mesh objects, since hierarchies and materials are probably not going to carry over in `.drc`. Typically if you need more than just geometry, you should use Draco compression inside a glTF file.

Mugen87 commented:
> BufferGeometry as input would be ideal – we're gradually converting all examples from Geometry to BufferGeometry.

Indeed. TBH, I'm against to add new code based on `THREE.Geometry`.

Apart from that, a `DRACOExporter` is a nice addition 👍

tentone committed:
Work in progress

tentone committed:
Support for buffer geometry

tentone committed:
Clean up

tentone commented:
Hi

Just updated the code to support booth BufferGeometry and the old nearly deprecated Geometry format, also added a couple of options to explicitly indicate if it needs to export uvs, normals and color.

Thanks a lot

Mugen87 commented on examples/js/exporters/DracoExporter.js:
```diff
@@ -0,0 +1,275 @@
+
+		} else if ( geometry.isGeometry === true ) {
+
+			var numFaces = geometry.faces.length;
```
Maybe it's easier to just convert an instance of `Geometry` to `BufferGeometry` and then use the same code path.

tentone replied:
Yeah it is a better approach makes it easier to maintain latter down the road and the code its doing pretty much the same as the built in conversion anyway.

I was trying to understand why my draco files were smaller for Geometry objects i was not exporting UV's propeties properly.

Will update the PR soon

tentone replied:
Updated the PR but i still need to change it to ensure that the buffer geometry is using faces, that we for the draco file.

Otherwise the encoding process will fail.

tentone replied:
Related #6324

donmccurdy replied:
Note that unfortunately Geometry—>BufferGeometry conversion has two inefficient side effects:

- vertex count increases significantly, even after toIndexed() call
- blank vertex colors are added to each vertex

The first can be fixed with BufferGeometryUtils.mergeVertices() but that’s probably better left to userland code. I’m not suggesting any change here (perhaps a warning on Geometry?), but just something to be aware of.

tentone committed:
Clean up

tentone committed:
Rollback

tentone committed:
Clean up

tentone committed:
Clean up calculate index

tentone commented:
Last version is completly woking for every scenarios i have tested. It is fully buffer geometry based.

From a performance standpoint it would be better to have dedicated Geometry code. I was having some trouble understanding how the UV are organizedin Geometry object so the UV were broken.

```
// Faces

var faces = new Uint32Array( geometry.faces.length * 3 );

for ( var i = 0; i < geometry.faces.length; i ++ ) {

	var index = i * 3;
	faces[ index ] = geometry.faces[ i ].a;
	faces[ index + 1 ] = geometry.faces[ i ].b;
	faces[ index + 2 ] = geometry.faces[ i ].c;

}

builder.AddFacesToMesh( mesh, geometry.faces.length, faces );

// Vertex

var vertices = new Float32Array( geometry.vertices.length * 3 );

for ( var i = 0; i < geometry.vertices.length; i ++ ) {

	var index = i * 3;
	vertices[ index ] = geometry.vertices[ i ].x;
	vertices[ index + 1 ] = geometry.vertices[ i ].y;
	vertices[ index + 2 ] = geometry.vertices[ i ].z;

}

builder.AddFloatAttributeToMesh( mesh, dracoEncoder.POSITION, geometry.vertices.length, 3, vertices );

// Normals

if ( options.exportNormals === true ) {

	var normals = new Float32Array( geometry.vertices.length * 3 );

	for ( var face of geometry.faces ) {

		normals[ face.a * 3 ] = face.vertexNormals[ 0 ].x;
		normals[ ( face.a * 3 ) + 1 ] = face.vertexNormals[ 0 ].y;
		normals[ ( face.a * 3 ) + 2 ] = face.vertexNormals[ 0 ].z;

		normals[ face.b * 3 ] = face.vertexNormals[ 1 ].x;
		normals[ ( face.b * 3 ) + 1 ] = face.vertexNormals[ 1 ].y;
		normals[ ( face.b * 3 ) + 2 ] = face.vertexNormals[ 1 ].z;

		normals[ face.c * 3 ] = face.vertexNormals[ 2 ].x;
		normals[ ( face.c * 3 ) + 1 ] = face.vertexNormals[ 2 ].y;
		normals[ ( face.c * 3 ) + 2 ] = face.vertexNormals[ 2 ].z;

	}

	builder.AddFloatAttributeToMesh( mesh, dracoEncoder.NORMAL, geometry.vertices.length, 3, normals );

}

// Uvs

if ( options.exportUvs === true ) {


	var uvs = new Uint32Array( geometry.faceVertexUvs.length * 6 );

	for ( var i = 0; i < geometry.faceVertexUvs.length; i ++ ) {

		var index = i * 6;
		var faceUvs = geometry.faceVertexUvs[ i ];

		uvs[ index ] = faceUvs[ 0 ].x;
		uvs[ index + 1 ] = faceUvs[ 0 ].y;

		uvs[ index + 2 ] = faceUvs[ 1 ].x;
		uvs[ index + 3 ] = faceUvs[ 1 ].y;

		uvs[ index + 4 ] = faceUvs[ 2 ].x;
		uvs[ index + 5 ] = faceUvs[ 2 ].y;

	}

	builder.AddFloatAttributeToMesh( mesh, dracoEncoder.TEX_COORD, geometry.vertices.length, 2, uvs );

}

// Color

if ( options.exportColor === true ) {

	var colors = new Uint32Array( geometry.colors.length * 3 );

	for ( var i = 0; i < geometry.colors.length; i ++ ) {

		var index = i * 3;
		colors[ index ] = geometry.colors[ i ].x;
		colors[ index + 1 ] = geometry.colors[ i ].y;
		colors[ index + 2 ] = geometry.colors[ i ].z;

	}

	builder.AddFloatAttributeToMesh( mesh, dracoEncoder.COLOR, geometry.vertices.length, 3, colors );

}
```

tentone commented:
I would like to advance with the current version if possible and later down the road i will try to clean up the Geometry code to improve performance when using those.

tentone committed:
Clean up (check quantization option)

tentone committed:
Clean up

tentone committed:
Clean up

mrdoob commented:
Thanks!
